src/template.rs

```diff
@@ -1,241 +1,375 @@
-    fn build(self: Box<Self>, world: &mut World, receipt: &mut Receipt);
+// -----------------------------------------------------------------------------
+// Callbacks and observers
+
+/// This is a helper for adding observers to a `template` macro.
+///
+/// ```
+/// # use i_cant_believe_its_not_bsn::*;
+/// # use bevy::prelude::*;
+/// template!{
+///     Name::new("MyEntity") => [
+///         on(|trigger: Trigger<Pointer<Click>>| {
+///             // Do something when "MyEntity" is clicked.
+///         });
+///         on(|trigger: Trigger<Pointer<Drag>>| {
+///             // Do something when "MyEntity" is dragged.
+///         });
+///     ];
+/// };
+/// ```
+pub fn on<E, B, M, I>(system: I) -> Callback
+where
+    E: Event,
+    B: Bundle,
+    I: IntoObserverSystem<E, B, M>,
+{
+    Callback::new(system)
 }
 
-/// Receipts contain hints about the previous outcome of building a particular
-/// prototype.
-#[derive(Default)]
-pub struct Receipt {
-    /// The entity this prototype was last built on (if any).
-    target: Option<Entity>,
-    /// The coponents it inserted.
-    components: HashSet<ComponentId>,
-    /// The receipts of all the children, organized by name.
-    children: HashMap<Anchor, Receipt>,
+#[derive(Component)]
+#[component(on_insert = insert_callback)]
+#[component(on_remove = remove_callback)]
+pub struct Callback {
+    observer: Option<Observer>,
 }
 
-/// A resource that tracks the receipts for root-level templates.
-#[derive(Resource, Default)]
-struct RootReceipt {
-    receipts: HashMap<Anchor, Receipt>,
+impl Callback {
+    pub fn new<E, B, M, I>(system: I) -> Callback
+    where
+        E: Event,
+        B: Bundle,
+        I: IntoObserverSystem<E, B, M>,
+    {
+        Callback {
+            observer: Some(Observer::new(system)),
+        }
+    }
 }
 
-/// A fragment is a tree of bundles with optional names. It implements
-/// [`Prototype`] and can be stored or used as a `Box<dyn Prototype>`.
-pub struct Fragment<B: Bundle> {
-    /// The name of the fragment, used to identify children across builds.
-    pub anchor: Option<String>,
-    /// The bundle to be inserted on the entity.
-    pub bundle: B,
-    /// The template for the children. This boils down to a type-errased
-    /// `Fragment` vector.
-    pub children: Template,
+impl From<Observer> for Callback {
+    fn from(observer: Observer) -> Callback {
+        Callback {
+            observer: Some(observer),
+        }
+    }
 }
 
-impl<B: Bundle> Prototype for Fragment<B> {
-    fn name(&self) -> Option<String> {
-        self.anchor.clone()
+fn insert_callback(mut world: DeferredWorld, entity_id: Entity, _component: ComponentId) {
+    let mut callback = world.get_mut::<Callback>(entity_id).unwrap();
+    let Some(mut observer) = mem::take(&mut callback.observer) else {
+        return;
+    };
+    if let Some(parent_id) = world.get::<Parent>(entity_id).map(|parent| parent.get()) {
+        observer.watch_entity(parent_id);
```

Comment:
Perhaps this is just obvious to other people (I'm not super familiar with how observers work), but perhaps this could do with a comment briefly explaining why we observe the parent if it's there, and why we fallback to listening to _all_ entities otherwise?

Reply:
Ah yeah, I'll be sure to document that. My thinking was, this could let you swap out different global observers by collecting them into a template.